brilirs/src/interp.rs

```diff
@@ -516,19 +622,58 @@ fn execute_value_op<T: std::io::Write>(
       let res = Value::Pointer(arg0.add(arg1));
       state.env.set(dest, res);
     }
+    Resolve => {
+      let res = match handlers.remove(&args[0]) {
+        Some(handle) => handle.0.join().unwrap(),
+        None => panic!("Unexpected execution error"),
+      };
+
+      state.env.set(dest, res.unwrap());
+    }
+    CompareAndSwap => {
+      //atomic
+      let arg0 = get_arg::<usize>(&state.env, 0, args);
+      //expected
+      let arg1 = get_arg::<i64>(&state.env, 1, args);
+      //update
+      let arg2 = get_arg::<i64>(&state.env, 2, args);
+      let res = state.atomics.compare_and_swap(arg0, arg1, arg2);
+      state.env.set(dest, Value::Int(res));
+    }
+    LoadAtomic => {
+      let arg0 = get_arg::<usize>(&state.env, 0, args) as usize;
+      let res = state.atomics.load_atomic(arg0);
+      state.env.set(dest, Value::Int(res));
+    }
+    SwapAtomic => {
+      let arg0 = get_arg::<usize>(&state.env, 0, args) as usize;
+      let arg1 = get_arg::<i64>(&state.env, 1, args);
+      let res = state.atomics.swap_atomic(arg0, arg1);
+      state.env.set(dest, Value::Int(res));
+    }
+    NewAtomic => {
+      let arg0 = get_arg::<i64>(&state.env, 0, args);
+      let atomics = Arc::into_raw(state.atomics.clone()) as *mut Atomics;
+      let res = unsafe { (*atomics).new_atomic(arg0) };
+      let _ = unsafe { Arc::from_raw(atomics) };
+
+      state.env.set(dest, res);
+    }
   }
   Ok(())
 }
 
-fn execute_effect_op<T: std::io::Write>(
-  state: &mut State<T>,
+fn execute_effect_op<T: std::io::Write + Sync + Send + 'static>(
+  state: &mut State,
   op: bril_rs::EffectOps,
   args: &[usize],
   funcs: &[usize],
   curr_block: &BasicBlock,
   // There are two output variables where values are stored to effect the loop execution.
   next_block_idx: &mut Option<usize>,
   result: &mut Option<Value>,
+  out: Arc<Mutex<T>>,
+  stop_flag: Option<Arc<AtomicBool>>,
```

Comment:
This stop_flag is somewhat mysterious to me. Why do we need to do this when the program ends up panic-ing anyways?

I think this could be implemented as `&Option<AtomicBool>`.

Reply:
Could this fit into State?